### PR TITLE
Add vault_kv2_delete module

### LINKED ISSUE #300: Add KV v2 delete functionality.
##### SUMMARY
I have a use case to be able to delete (versions) of kv2 data in Vault. I may be able to contribute to this feature.

Since the KV v2 API is a bit more complex than some of the other secret engines, I'm not sure if it would make sense to have `vault_kv2_{verb}` modules, or just generic `vault_{rest_api_verb}` modules (i.e. futureproofing for other features like list, metadata, destroy and undelete). 

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
- vault_kv2_delete_version ?
- vault_delete ?

##### ADDITIONAL INFORMATION

- If `versions` not supplied, will default to latest version (same default as api)

```yaml
- name: Delete version 5 of a secret from kv2 with a different mount via the remote host
  community.hashi_vault.vault_kv2_delete_version:
    url: https://vault:8201
    path: path/to/secret
    versions: [5]

```

### PR DESCRIPTION
**Pull request by idwagner**
##### SUMMARY
Implementation of vault_kv2_delete module (hvac delete_latest_version_of_secret)

Fixes #300

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

vault_kv2_delete

##### ADDITIONAL INFORMATION

This implements the delete latest version functionality of a KV v2 secret in Vault. This is a WIP and is missing:
- Module docs
- Integration tests

### CODE REVIEW

File: plugins/modules/vault_kv2_delete.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    try:
+        module.authenticator.validate()
+        module.authenticator.authenticate(client)
+    except (NotImplementedError, HashiVaultValueError) as e:
+        module.fail_json(msg=to_native(e), exception=traceback.format_exc())
+
+    try:
+        raw = client.secrets.kv.v2.delete_latest_version_of_secret(path=path, mount_point=engine_mount_point)
+    except hvac.exceptions.Forbidden as e:
```
Comment (briantist) [inline plugins/modules/vault_kv2_delete.py:61]: You may want to check for some more exceptions, depending on what `hvac` might raise/handle (best to check the source there to get an idea what to handle here).

File: tests/unit/plugins/modules/test_vault_kv2_delete.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2022 Brian Scholer (@briantist)
```
Comment (briantist) [inline tests/unit/plugins/modules/test_vault_kv2_delete.py:2]: I know it's a WIP but update this with your info.

And the method names will need to change from the kv2_get names as well.

**Review by briantist — COMMENTED** [review]
Thank you @idwagner ! This is a great start :)

File: tests/unit/plugins/modules/test_vault_kv2_delete.py (hunk 2 above)
Comment (idwagner) [reply tests/unit/plugins/modules/test_vault_kv2_delete.py:2]: done :)

File: plugins/modules/vault_kv2_delete.py (hunk 1 above)
Comment (idwagner) [reply plugins/modules/vault_kv2_delete.py:116]: Looks like hvac does not do any error handling for this. I also peeked at the Vault source, and it doesn't do anything special in the delete itself, just passing through any errors it gets such as any errors updating the metadata. If theres anything else you can think of, I'd be happy to add it.

File: plugins/modules/vault_kv2_delete.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        module.fail_json(msg="Forbidden: Permission Denied to path ['%s']." % path, exception=traceback.format_exc())
+
+    module.exit_json(changed=True, response_code=raw.status_code)
```
[the remark this answers is not in the document]
Comment (idwagner) [reply plugins/modules/vault_kv2_delete.py]: I've added the vault_write logic along with the associated pertinent docs on the return.

File: plugins/modules/vault_kv2_delete.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    HVAC_IMPORT_ERROR = traceback.format_exc()
+else:
+    HVAC_IMPORT_ERROR = None
+    HAS_HVAC = True
+
+def run_module():
+
+
+    argspec = HashiVaultModule.generate_argspec(
+        engine_mount_point=dict(type='str', default='kv'),
```
[the remark this answers is not in the document]
Comment (idwagner) [reply plugins/modules/vault_kv2_delete.py]: done :)

**Comment by idwagner** [comment]
@briantist few questions,

Is there any convention for adding policies to the integration tests by path or key? Should I create a policy on a new kv path for these tests rather than modifying an existing path permission?

Also, I'm thinking maybe I should rename `vault_kv2_delete` to `vault_kv2_delete_latest_version` to more closely match hvac's methods. Further development could have modules like `vault_kv2_delete_version` `vault_kv2_undelete` and `vault_kv2_destroy` etc...

**Comment by briantist** [comment]
> @briantist few questions,
> 
> Is there any convention for adding policies to the integration tests by path or key? 

Yes, have a look at [this file for the tasks](https://github.com/ansible-collections/community.hashi_vault/blob/main/tests/integration/targets/setup_vault_configure/tasks/configure.yml), and [this file for the policy definitions](https://github.com/ansible-collections/community.hashi_vault/blob/main/tests/integration/targets/setup_vault_configure/vars/main.yml).

> Should I create a policy on a new kv path for these tests rather than modifying an existing path permission?

Feel it out and see what makes sense. A new path is probably the way to go, where the "setup" parts of the integration tests would create the secret(s) to be deleted in the tests.

Also have a look at the [test plugins](https://github.com/ansible-collections/community.hashi_vault/tree/main/tests/integration/targets/setup_vault_test_plugins), in case a new one needs to be added there for any reason (for example for creating/deleting the secrets).

> Also, I'm thinking maybe I should rename `vault_kv2_delete` to `vault_kv2_delete_latest_version` to more closely match hvac's methods. Further development could have modules like `vault_kv2_delete_version` `vault_kv2_undelete` and `vault_kv2_destroy` etc...

I think I'm still leaning toward a single `vaukt_kv2_delete` that will have the ability to optionally delete a specific version or versions depending on how the options are set. There can still be other modules like `vault_kv2_undelete` and `vault_kv2_destroy`. Any reason in particular you're thinking of going back the other way?

File: tests/unit/plugins/modules/test_vault_kv2_delete.py (hunk 2 above)
Comment (briantist) [reply tests/unit/plugins/modules/test_vault_kv2_delete.py:2]: Method names look good, but it's still my name in the copyright ;)

**Comment by idwagner** [comment]
@briantist I believe this is about ready with the integration tests in place. Two of the tests failed with errors that lead me to believe they are re-using a vault session from a previous test? I setup some seed data in `setup_vault_configure` assuming that would be run during each integration test, however the failed tests don't match that initial seed data.

https://github.com/idwagner/community.hashi_vault/blob/a0f8e7b4597408e8ad330a612e4a783d9b3db6a1/tests/integration/targets/setup_vault_configure/tasks/configure.yml#L100-L107

**Comment by briantist** [comment]
> @briantist I believe this is about ready with the integration tests in place. Two of the tests failed with errors that lead me to believe they are re-using a vault session from a previous test? I setup some seed data in `setup_vault_configure` assuming that would be run during each integration test, however the failed tests don't match that initial seed data.
> 
> https://github.com/idwagner/community.hashi_vault/blob/a0f8e7b4597408e8ad330a612e4a783d9b3db6a1/tests/integration/targets/setup_vault_configure/tasks/configure.yml#L100-L107

Ah! I'm sorry you hit that error, but I am happy to see that those tests are failing in exactly the right situation they were designed to!

> Two of the tests failed with errors that lead me to believe they are re-using a vault session from a previous test?

They are doing that, intentionally. The key is that I want the tests to be able to run against a local instance of Vault that doesn't need setup on every single run, because it speeds up local iteration.

So in your case, the problem, is that after you create those secret versions and delete some of them, those versions do not exist again. And if you tried to insert new versions, they would not "fill in" those missing ones either they would continue adding new versions to the end (versions 6, 7, 8, etc.).

For a test like this, instead of setting up that secret in the "test-wide" re-usable secrets, you should set it up in `tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_setup.yml` in the configuration tasks. In here, you should:
1. Delete the existing secret completely if it exists (this must succeed whether it exists or not)
2. Create the secret and all of its versions.

It should be removed from the general vault configure target.

File: galaxy.yml [hunk 5]
```diff
@@ -7,6 +7,7 @@ readme: README.md
 authors:
   - Julie Davila (@juliedavila) <julie(at)davila.io>
   - Brian Scholer (@briantist)
+  - Isaac Wagner (@idwagner)
```
[the remark this answers is not in the document]
Comment (idwagner) [reply galaxy.yml:10]: Nor worries, I will remove that 👍

**Comment by idwagner** [comment]
It looks like the change corrected one of the tests, but other timed out. Can that test be re-run without a push?

File: plugins/modules/vault_kv2_delete.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+    argspec = HashiVaultModule.generate_argspec(
+        engine_mount_point=dict(type='str', default='secret'),
+        path=dict(type='str', required=True),
+        versions=dict(type='list', elements='int', required=False)
+    )
+
+    module = HashiVaultModule(
+        argument_spec=argspec,
+        supports_check_mode=True
```
Comment (briantist) [inline plugins/modules/vault_kv2_delete.py:105]: At the moment, the module does not account for check mode, and so with this set to `True`, check mode would delete secrets 😱 

One issue currently (same reason status is always `changed`) is that without checking the current status of the secret, check mode has nothing to do but return `changed`.

I'm mulling over the idea of using metadata reads to get current versions status/secret existence to be able to solve both of those, but their use should be optional (a client may have permission to delete but not to hit those other endpoints).

I'll look at this again tomorrow and think about whether that makes sense at all, and if so whether it should be in this PR or a follow-up.

The first thing we should do is add an integration test and unit test ([example](https://github.com/ansible-collections/community.hashi_vault/blob/main/tests/unit/plugins/modules/test_vault_login.py#L98)) for check mode before making any additional changes, and those tests should fail right now.

File: tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_setup.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        mount_point: '{{ vault_kv2_mount_point }}'
+        path: '{{ vault_kv2_versioned_path }}/secret6'
+
+    - name: Set up a multi versioned secret for delete (v2)
+      vault_ci_kv_put:
+        version: 2
+        mount_point: '{{ vault_kv2_mount_point }}'
+        path: '{{ vault_kv2_versioned_path }}/secret6'
+        secret:
+          v: value{{item}}
```
Comment (briantist) [inline tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_setup.yml:28]:
```suggestion
          v: value{{ item }}
```

**Review by briantist — COMMENTED** [review]
See the comment about check mode.

The rest are mostly small changes.

In case you haven't seen it before, you can commit multiple suggestions from review at once, if you view the comments in the Files tab. Then, there will be an option to add to batch, and then you can commit the batch at once.

File: plugins/modules/vault_kv2_delete.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+short_description: Delete the latest version of a secret from HashiCorp Vault's KV version 2 secret store
+requirements:
+  - C(hvac) (L(Python library,https://hvac.readthedocs.io/en/stable/overview.html))
+  - For detailed requirements, see R(the collection requirements page,ansible_collections.community.hashi_vault.docsite.user_guide.requirements).
+description:
+  - Deletes the latest version of a secret from HashiCorp Vault's KV version 2 secret store.
+notes:
+  - This module always reports C(changed) status because it cannot guarantee idempotence.
+  - Use C(changed_when) to control that in cases where the operation is known to not change state.
+seealso:
```
[the remark this answers is not in the document]
Comment (briantist) [reply plugins/modules/vault_kv2_delete.py:23]: My original suggestion was to link to `vault_kv2_get`, then you changed it to `vault_kv2_delete` in https://github.com/ansible-collections/community.hashi_vault/pull/304/commits/9cafc720ed62d9dd923b29460b9743d21c33d292 , then deleted it in https://github.com/ansible-collections/community.hashi_vault/pull/304/commits/1b3aeb181a03610759d0d2ea738034fbe100eb6c as being self-referential 😅

I still think a link to `vault_kv2_get` is a good idea though. I find it's nice to link to other kv2 modules, do you prefer not having it?

```suggestion
seealso:
  - module: vault_kv2_get
```

File: plugins/modules/vault_kv2_delete.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+  - Isaac Wagner (@idwagner)
+short_description: Delete one or more versions of a secret from HashiCorp Vault's KV version 2 secret store
+requirements:
+  - C(hvac) (L(Python library,https://hvac.readthedocs.io/en/stable/overview.html))
+  - For detailed requirements, see R(the collection requirements page,ansible_collections.community.hashi_vault.docsite.user_guide.requirements).
+description:
+  - Delete one or more versions of a secret from HashiCorp Vault's KV version 2 secret store.
+notes:
+  - This module always reports C(changed) status because it cannot guarantee idempotence.
+  - Use C(changed_when) to control that in cases where the operation is known to not change state.
```
Comment (briantist) [inline plugins/modules/vault_kv2_delete.py:22]:
```suggestion
  - Use C(changed_when) to control that in cases where the operation is known to not change state.
  - In check mode, the module returns C(changed) status without contacting Vault. Consider using M(community.hashi_vault.vault_kv2_get) to verify the existence of the secret first.
```

File: plugins/modules/vault_kv2_delete.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    path: secret/mysecret
+    versions: [1, 3]
+    auth_method: userpass
+    username: user
+    password: '{{ passwd }}'
+"""
+
+RETURN = """
+data:
+  description: The raw result of the delete against the given path.
```
Comment (briantist) [inline plugins/modules/vault_kv2_delete.py:72]:
```suggestion
  description: 
    - The raw result of the delete against the given path.
    - This is usually empty, but may contain warnings or other information.
```

File: plugins/modules/vault_kv2_delete.py (hunk 6 above)
Comment (briantist) [reply plugins/modules/vault_kv2_delete.py:105]: Thank you, I think for now the empty result with no Vault connection is fine.

The unit tests are great! Please also add an integration test for check mode; it should be easy, just checking that the result `is changed` and that the deletion did not happen.

Even though the unit test covers this a little more thoroughly, it's always good to have it covered by integration too just to have the code actually invoked by ansible.

File: tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_setup.yml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      vault_ci_token_create:
+        policies: [test-policy]
+      register: user_token_cmd
+
+- name: Configuration tasks
+  module_defaults:
+    vault_ci_kv2_destroy_all: '{{ vault_plugins_module_defaults_common }}'
+    vault_ci_kv_put: '{{ vault_plugins_module_defaults_common }}'
+  block:
+
```
Comment (briantist) [inline tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_setup.yml:16]: nit
```suggestion
```

File: tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_test.yml [hunk 12]
```diff
@@ -0,0 +1,175 @@
+---
+
```
Comment (briantist) [inline tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_test.yml:2]: nit
```suggestion
```

File: plugins/modules/vault_kv2_delete.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (idwagner) [reply plugins/modules/vault_kv2_delete.py:23]: I don't know what happened there 😃 I agree `vault_kv2_get` should be referenced.

File: tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_test.yml [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+              - "kv2_result['result']['data']['versions']['4']['deletion_time'] is search(regex_secret_version_not_deleted)"
+              - "kv2_result['result']['data']['versions']['5']['deletion_time'] is search(regex_secret_version_not_deleted)"
+            fail_msg: 'Test Seed value did not contain expected data.'
+
+        - name: Try kv2 delete latest version in check mode
+          register: kv2_result
+          community.hashi_vault.vault_kv2_delete:
+            path: "{{ vault_kv2_versioned_path }}/secret6"
+          check_mode: true
+
```
Comment (briantist) [inline tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_test.yml:60]: let's add assertions for the output result of the deletions too; the intention will be to test these after each call to the delete module.

```suggestion

        - assert:
            that: &delete_assertions
              - kv2_result is changed
              - kv2_result.data == {}

```

File: tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_test.yml [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+            fail_msg: 'Secret version was deleted while in check mode.'
+
+        - name: Try kv2 delete version 1 and 3
+          register: kv2_result
+          community.hashi_vault.vault_kv2_delete:
+            path: "{{ vault_kv2_versioned_path }}/secret6"
+            versions:
+              - 1
+              - 3
+        - assert:
```
Comment (briantist) [inline tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_test.yml:108]: spacing is a nit

But something is strange here. I don't understand how this assertion is passing. `kv2_result` should contain nothing but `changed=true` and `data={}`, so how are the basics assertions not failing?

```suggestion

        - assert:
```

Comment (briantist) [reply tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_test.yml:108]: Ah! I have figured this out. We cannot use YAML anchors like this.

With the following:
```yaml
- assert:
    that: &a
      - 1
      - 2
```

The anchor is referring to the entire list value: `[1, 2]`.

When using it like this:
```yaml
- assert:
    that:
      - *a
      - b
```

Then, what we have for the list of assertions then is actually this: `[[1,2], "b"]`

So the first list item is a non-empty list which always evaluates to true.

For dictionaries, YAML anchors have the merge operator `<<` but there is no such thing for list.

We don't have a lot of good options for this. Either duplicate all of the "basic" assertions, or use vars in a way like this:

```
- name: some block
  vars:
    basic_assertions: &basics
      - 1
      - 2
 
- name: simple case - basics only - option 1
  assert:
    that: *basics

- name: simple case - basics only - option 2
  assert:
    that: '{{ basic_assertions }}'

- name: complex 1 - literals
  assert:
    that: "{{ basic_assertions + ['condition 1', 'condition 2'] }}"

- name: complex 2 - vars
  vars:
    these_assertions:
      - condition 1
      - condition 2
  assert:
    that: '{{ basic_assertions + these_assertions }}'
```

**Comment by briantist** [comment]
There is one more thing I forgot (#252 🤦): please add this module to the action group:
- https://github.com/ansible-collections/community.hashi_vault/blob/main/meta/runtime.yml#L4

File: tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_test.yml (hunk 14 above)
Comment (idwagner) [reply tests/integration/targets/module_vault_kv2_delete/tasks/module_vault_kv2_delete_test.yml:113]: I feel its a lot more readable to just duplicate the lines from the anchor where its needed (which is only a handful).. Cleaning up some of that now and will push something up once i run some local tests

**Review by briantist — APPROVED** [review]

**Comment by briantist** [comment]
@idwagner thanks so much for this contribution! It's now released in [version 3.4.0](https://github.com/ansible-collections/community.hashi_vault/releases/tag/3.4.0)! 🥳

**Pull request merged**